File: py/leetcode-300/main.py
from typing import List
from sortedcontainers import SortedList

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)

        x = SortedList()
        x.add((nums[0], 1))
        result = 1
        for i in range(1, n):
            v = nums[i]
            z = 1
            j = x.bisect_right((v-1, n+1))
            if j > 0:
                z = x[j-1][1] + 1
            
            # since the lengths are stricly increasing
            # if we need to pop an element it has to be x[i]
            if j < len(x):
                if x[j][1] <= z:
                    x.pop(j)

            x.add((v, z))
            print(x)
            if z > result:
                result = z

        return result

def main():
    sol = Solution()

    nums = [10,9,2,5,3,7,101,18]
    print(sol.lengthOfLIS(nums))

    nums = [0,1,0,3,2,3]
    print(sol.lengthOfLIS(nums))

    nums = [7,7,7,7,7,7,7]
    print(sol.lengthOfLIS(nums))

if __name__ == "__main__":
    main()

File: go/leetcode-300/main.go
package main

func lengthOfLIS(nums []int) int {
	// longest strictly increasing subsequence
	// O(n^2) is easy
	// for each i compute d(i) = longest subsequence ending at index i
	// look at each j < i to use as previous element

	// return simpleLIS(nums)

	// but we can do better
	// suppose the longest sequence so far ends in value v and has length l
	// and at index i we have value w and length k
	// if v < w then we can use that sequence to get a higher value
	// if v >= w there are two cases
	//    l =< k, i.e. then we don't need (v, l), since any later subsequence
	//    	can just use (w, k)
	//    l > k keep (v,l)
	// the point is that this will form an increasing sequence in both
	// the value and length of sequence, and thus we can do binary search
	// just need to keep these as a sorted list
	// in summary: when we add (w, k) we can remove all (v, l)
	// with v >= w and l <= k
	// but every index will be added/removed at most once

	return betterLIS(nums)
}

func simpleLIS(nums []int) int {
	n := len(nums)
	d := make([]int, n)
	d[0] = 1
	result := 1
	for i := 1; i < n; i++ {
		v := 1
		for j := 0; j < i; j++ {
			if nums[j] < nums[i] && d[j]+1 > v {
				v = d[j] + 1
			}
		}
		d[i] = v
		if v > result {
			result = v
		}
	}
	return result
}

func betterLIS(nums []int) int {
	return 0
}
